ui-support-service/src/main/java/org/dcsa/uisupport/service/UiTransportCallService.java

```diff
@@ -0,0 +1,114 @@
+    } else if (transportCallTO.vessel() == null || transportCallTO.vessel().vesselIMONumber() == null) {
+      throw ConcreteRequestErrorMessageException.invalidInput("Cannot create transport call where vessel.vesselIMONumber is missing");
+    } else if (transportCallTO.facilityCode() == null ^ transportCallTO.facilityCodeListProvider() == null) {
+      if (transportCallTO.facilityCode() == null) {
+        throw ConcreteRequestErrorMessageException.invalidInput("Cannot create transport call where facility code list provider is present but facility code is missing");
+      } else {
+        throw ConcreteRequestErrorMessageException.invalidInput("Cannot create transport call where facility code is present but facility code list provider is missing");
+      }
+    }
+
+    Location location = locationRepository.save(
+      Location.builder()
+        .id(UUID.randomUUID().toString())
+        .UNLocationCode(transportCallTO.UNLocationCode())
+        .facility(findFacility(transportCallTO).orElse(null))
+        .build()
+    );
+    org.dcsa.jit.persistence.entity.Service service =
+      serviceService.ensureServiceExistsByCarrierServiceCode(transportCallTO.carrierServiceCode());
+
+    TransportCall entityToSave = TransportCall.builder()
+      .transportCallReference(UUID.randomUUID().toString())
+      .transportCallSequenceNumber(Objects.requireNonNullElse(transportCallTO.transportCallSequenceNumber(), 1))
+      .facility(null) // Go through location to find facility,
+      .facilityTypeCode(enumMappers.facilityTypeCodeToDao(transportCallTO.facilityTypeCode()))
+      .location(location)
+      .modeOfTransportCode(enumMappers.modeOfTransportToDao(transportCallTO.modeOfTransport()).getCode().toString())
+      .vessel(vesselService.ensureVesselExistsByImoNumber(transportCallTO.vessel().vesselIMONumber()))
+      .importVoyage(Voyage.builder().carrierVoyageNumber(transportCallTO.importVoyageNumber()).service(service).build())
+      .exportVoyage(Voyage.builder().carrierVoyageNumber(transportCallTO.exportVoyageNumber()).service(service).build())
+      .portCallStatusCode(null)
+      .build();
+
+    TransportCall saved = transportCallRepository.save(entityToSave);
+
+    return TransportCallWithTimestampsTO.builder()
+      .transportCallID(saved.getId()) // Just adding the id for easier testing
+      .transportCallTO(transportCallMapper.toTO(saved))
+      .build();
+  }
+
+  private Optional<Facility> findFacility(TransportCallTO transportCallTO) {
+    if (transportCallTO.facilityCodeListProvider() != null) {
+      return switch (transportCallTO.facilityCodeListProvider()) {
+        case SMDG ->
+          facilityRepository.findByUNLocationCodeAndFacilitySMDGCode(transportCallTO.UNLocationCode(), transportCallTO.facilityCode());
+        case BIC ->
+          facilityRepository.findByUNLocationCodeAndFacilityBICCode(transportCallTO.UNLocationCode(), transportCallTO.facilityCode());
```

Comment:
Does this ensure a (readable) error message if we do not find the facility?

Reply:
It should lead to a 400 Bad Request (or thereabouts) if a code is present and we cannot resolve it rather than a "silent null corruption".